windows/src/developer/TIKE/xml/layoutbuilder/builder-charmap.js

```diff
@@ -0,0 +1,188 @@
+//
+// Character map drag+drop and double-click insertion
+//
+
+builder.dragDrop = {};
+(function(builder, dragDrop) {
+  dragDrop.isSubKey = function(key) {
+    return key.parent().attr('id') == 'sk';
+  };
+
+  dragDrop.getKeyElementFromElement = function(elem) {
+    if(elem.classList.contains('key')) {
+      return elem;
+    }
+
+    if(!elem.parentElement) {
+      return null;
+    }
+
+    if(elem.parentElement.classList.contains('key')) {
+      return elem.parentElement;
+    }
+
+    return null;
+  };
+
+  dragDrop.getKeyElementFromDragTarget = function(target) {
+    let key = dragDrop.getKeyElementFromElement(target);
+    let isIdField = false;
+
+    if(key) {
+      key = $(key);
+    } else {
+      if(target.id == 'inpKeyCap') {
+        key = builder.selectedKey();
+      } else if(target.id == 'inpKeyName') {
+        key = builder.selectedKey();
+        isIdField = true;
+      } else if(target.id == 'inpSubKeyCap') {
+        key = builder.selectedSubKey();
+      } else if(target.id == 'inpSubKeyName') {
+        key = builder.selectedSubKey();
+        isIdField = true;
+      }
+
+      if(!key) {
+        return null;
+      }
+    }
+
+    return { key: key, isIdField: isIdField };
+  };
+
+  dragDrop.isHighSurrogate = function(character) {
+    let codeUnit = character.charCodeAt(0);
+    return codeUnit >= 0xD800 && codeUnit <= 0xDBFF;
+  };
+
+  dragDrop.isLowSurrogate = function(character) {
+    let codeUnit = character.charCodeAt(0);
+    return codeUnit >= 0xDC00 && codeUnit <= 0xDFFF;
+  };
+
+  dragDrop.surrogatePairToValue = function(character) {
```

Comment:
Does Windows have any shared utility code for handling surrogate pairs?

Reply:
Well, this is JS code. It's currently completely isolated from all the other JS in Keyman. Not keen on refactoring to that level at this point. If necessary, we can add a TODO to move shared code around, but it's not going to win us much...

Reply:
And ... that surrogate pair code is all gone now! Yay!